gateway/gateway-requests/src/registration/handshake/gateway.rs

```diff
@@ -49,11 +50,28 @@ impl<'a> GatewayHandshake<'a> {
                 }
 
                 // init: <- pub_key || g^x
-                let (remote_identity, remote_ephemeral_key) = check_processing_error(
+                let init_message = check_processing_error(
                     State::<S>::parse_init_message(received_init_payload),
                     &mut state,
                 )
                 .await?;
+
+                let credential = init_message.credential();
+
+                check_processing_error(
+                    {
+                        if !credential.verify(validator_urls) {
```

Comment:
This is the issue I described before, we really don't want to be talking with all validators here every single time.... Just use existing verification key.
Also, should the coconut verification be happening immediately? Out of all the steps in the handshake this is **the** most expensive step, so perhaps it should be left at the very end? Though I guess we already have the credential on hand so maybe not... (this is just some thinking aloud, I'm not sure what's the correct answer to that)

Reply:
Fixed by making it async